src/diagnostic/mod.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright 2024 FastLabs Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Markers to enrich log records with additional information.
+
+use std::borrow::Cow;
+
+#[cfg(feature = "fastrace")]
+pub use self::fastrace::FastraceDiagnostic;
+pub use self::thread_local::ThreadLocalDiagnostic;
+
+#[cfg(feature = "fastrace")]
+mod fastrace;
+mod thread_local;
+
+/// A visitor to walk through diagnostic key-value pairs.
+pub trait Visitor {
+    /// Visits a key-value pair.
+    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
+    where
+        K: Into<Cow<'k, str>>,
+        V: Into<Cow<'v, str>>;
+}
+
+/// Mapped Diagnostic Context (MDC).
+///
+/// A lighter technique consists of uniquely stamping each log request.
+#[derive(Debug)]
+pub enum Diagnostic {
+    #[cfg(feature = "fastrace")]
+    Fastrace(FastraceDiagnostic),
+    ThreadLocal(ThreadLocalDiagnostic),
+}
+
+impl Diagnostic {
+    pub fn name(&self) -> &'static str {
```

Comment:
What is this method for? Since it's supposed to by dyn.

Reply:
Previously I use it to filter all the context provided from FastraceDiagnostic in some case.

That is, a key-value in MDC can be considered as (context_name, key_name, value) and implementor of append/layout can filter kvs by either name. But it may not be used now and perhaps over design.

Reply:
Removed in a new PR.